### PR TITLE
Preliminary fixes to allow perfectly tiled generation

### PR DESCRIPTION
**Pull request by indyteo**
### Type of modification

- Code
  - General
  - Outputs
- Tests

### Changes

Preliminary fixes to allow perfectly tiled generation.

#### Feature description

While experimenting with tiled generation, I ran into issues due to imperfect tiling (misaligned edges, gaps between tiles...) and this PR fixes all those little bugs.

Three small fixes are included:
1. (Minecraft only) The generated world was offset by 1 along the Z axis.
2. The envelope computed by the `Generation` class was 1 unit short compared to the generated world size.
3. The world spawn was not located at the center of the generated world, but rather at fixed location (0, 0).

I will quickly describe each issue as well as why it was present and how I fixed it.

**1. Z offset in Minecraft**
In Minecraft, the Z axis is facing south, as opposed to both Minetest and IRL maps where Z is traditionally facing north. Initially, we didn't notice this until I start to experiment with roads (now in PR #58). At the early stage of development, I committed a fix via PR #50, to use `-z` instead. However, what I failed to notice after this fix was the 1-block-offset I introduced!

To understand this bug, we first need to take a closer look at the world's size (when even), and how the integer coordinates are handled (explained here in 2d for simplicity):
- Voxel position (0, 0) is the voxel from decimal (0.0, 0.0) to (0.99, 0.99).
- The absolute middle of the generated world is decimal (0.0, 0.0), meaning that if we split the world into 4 equals quarters, voxel (0, 0) belongs to the upper-right (IRL north-east) quarter.
- With world size (32, 32), the resulting limits are voxel (-16, -16) to voxel (15, 15).

With that said, it becomes clear that a simple minus sign in front of the `z` value won't work. A voxel on the border of the world with Z = 15 will be placed at Z = -15 (instead of -16), and voxel at Z = -16 will be placed at Z = 16 (instead of 15), causing a +1 shift of the whole generated world. Furthermore, the limits are also affected, meaning that they become -15 to 16.

Luckily enough, the fix is pretty easy: Subtract 1 to Z when converting from "Generator => Minecraft", and add 1 back to reverse from "Minecraft => Generator".

However, this fix will cause the shift to be present on odd world size. Nevertheless, odd world size is unusual, as Minecraft stores blocks (voxels) in 16x16x16 areas, and tiling can only be achieved with size divisible by 512, thus I prefer to focus on even size.

**2. Envelope too small**
The `Generation` class computes an `Envelope` from the in-game world size, by applying the appropriate transformation. The envelope is the smallest bounding box in decimal coordinates that contains all the generated world. However, I noticed that the computed envelope was 1 unit short on each axis, effectively causing the last row and the last column of voxels to be missing data.

The reason behind this bug is somewhat similar to the previous one, it depends on the interpretation of the world's limits. As said before, size (32, 32) leads to bounds (-16, -16) and (15, 15). In voxel coordinates, this is exact (because the voxel from 15.0 to 15.99 is included), however, in decimal coordinates, this leads to a size of 31, excluding the last voxel.

The fix for this is trivial, to include in the envelope the last row/column of voxels, we just need to take decimal bounds from voxel min to voxel max + 1 (so the last voxel is inside).

**3. Spawn location at (0, 0)**
The in-game spawn location is computed inside `SampleImplementation`. The current code was using hardcoded (0, 0) coordinates instead of the real center (computed from world's limits). Furthermore, the height was retrieved from the ground heightmap without even checking that those coordinates were in-bounds.

When trying to introduce an "offset" to the generation, I moved the world's limits and at some point, the location (0, 0) was outside of the generated area. This caused an exception when trying to read the height.

The fix is also relatively simple, the spawn location is now computed as the center of the generated world area, and the height is assured to be valid. This does not change anything without an offset, and because the offset mechanism is not included in this PR, this change effectively has not effect on current generation.

#### Showcase

I will illustrate the issues using a 32x32 Minecraft generation.

_From above:_
minecraft-32x32-bugged-above
In yellow, the world border is also offset by 0.5. this issue was not mentioned because it was fixed as a result of Minecraft Z offset issue and small envelope issue.
In red, the offset of the generated world along the Z axis (red arrow). Voxels on the right are missing, because the whole map is shifted by 1 and extra voxels goes beyond the expected border (on the left).

_From beneath:_
minecraft-32x32-bugged-below
In green, we can see a row and column of voxels at a default height of 0, because no height data was loaded for them (envelope too small, they were outside).
In purple, the correct height where they should have been generated.

_With fixes applied:_
minecraft-32x32-fixed
The generated world now perfectly fills the expected area

_Note: The outside of the border is randomly generated by Minecraft the first time the world is opened in the game. This also explains the difference of grass color and flowing liquids (will probably need to be addressed in a later PR)_

### Reason

While those bugs were not a big deal with current generation (only caused imperfection at the edge of the generated world), they needed to be addressed before starting working on tiling. In fact, the edge of generated tiles must be perfect to align properly each other!

### Self-checks

- [x] The code has unit tests associated
- [x] The code has Javadoc Comments associated
- ~~Complex / Unexpected code is explained / justified with a small comment~~
- ~~Relevant documentation inside the `/docs` folder has been updated~~
- [x] All examples in `docs/usage/Examples.md` work the same (or have been adapted if subject to changes in this PR)
- [x] Git history is clean (each commit accomplish a single task and describe it accordingly)
- [x] The texts have been proofread (documentation, error messages, logs, comments...)

### CODE REVIEW

File: src/main/java/com/ignfab/minalac/generator/SampleImplementation.java [hunk 1]
```diff
@@ -72,7 +73,10 @@ public static void main(String[] args) throws FactoryException, InterruptedExcep
 
         System.out.println("Saving world");
         VoxelWorldMetadata metadata = generation.world().getMetadata();
-        metadata.setSpawn(new WorldCoords3d(0, 0, generation.heightmaps().get("ground").get(0, 0) + 1));
+        WorldBBox3d limits = generation.world().limits();
+        int spawnX = (limits.minX() + limits.maxX() + 1) / 2;
+        int spawnY = (limits.minY() + limits.maxY() + 1) / 2;
+        metadata.setSpawn(new WorldCoords3d(spawnX, spawnY, generation.heightmaps().get("ground").get(spawnX, spawnY) + 1));
```
Comment (pyrollo) [inline src/main/java/com/ignfab/minalac/generator/SampleImplementation.java:79]: Plusieurs réflexions me viennent.
La première c'est qu'on pourrait mettre les coordonnées du spawn dans le yaml, comme ça on ne s'inquiète plus de ce calcul.
La seconde c'est que, vu le paramétrage, le centre du monde généré est toujours 0,0 puisqu'on défini le monde par rapport à ce point, avec les étendues horizonales.

Comment (indyteo) [reply src/main/java/com/ignfab/minalac/generator/SampleImplementation.java:79]: Comme je l'ai expliqué, c'est pour préparer le tiling, donc oui en l'état ça ne change rien (le calcul donne (0, 0) avec le paramétrage possible actuel), mais grâce à cette modif, sans rien changer d'autre qu'un léger offset dans la classe `Generation`, on peut générer des tuiles différentes / centrer sur un point différent

(D'autant plus que le calcul certes donne (0, 0), mais est + souple qu'une valeur hardcodée, et on est dans le SampleImplementation, ça reste brouillon / expérimental avant qu'on l'intègre pleinement dans une partie du code)

Comment (pyrollo) [reply src/main/java/com/ignfab/minalac/generator/SampleImplementation.java:79]: Ici ce que je ne capte pas c'est le rapport entre le spawn et le tuilage. Pour moi, quelque soit la tuile générée, le spawn reste au même point (et de toutes façons, il ne devrait pas être choisi dans les tuiles, mais au racommodage final). Ou alors tu veux pouvoir générer des tuiles jouables indépendemment ?

Comment (pyrollo) [reply src/main/java/com/ignfab/minalac/generator/SampleImplementation.java:79]: Je garde sous le coude l'idée de pouvoir placer le spawn à un autre endroit que le centre, via le Yaml, je crois que ça peut être utile.

Comment (indyteo) [reply src/main/java/com/ignfab/minalac/generator/SampleImplementation.java:79]: Je suis bien d'accord que le spawn (et les autres fichiers de méta-données du monde) ne devrait pas être généré pour chaque tuile, cependant dans un premier temps ça reste le plus simple, et le code revient au même dans le reste des cas, donc je ne vois pas vraiment le problème... En gros, c'est identique actuellement et plus pratique pour la suite.

Et oui du coup on peut ne générer qu'une seule tuile indépendante, et avoir les méta-données du monde (spawn et tout ce qui le rend jouable) n'est pas inintéressant puisque ça permet de débugger individuellement des tuiles (ça m'a permis de tester mon tuilage, après je suis d'accord que ça devrait plus forcément être nécessaire, mais pourquoi pas)

Comment (indyteo) [reply src/main/java/com/ignfab/minalac/generator/SampleImplementation.java:79]: La raison principale pour laquelle j'ai fait cette modification c'est pour me débarrasser de l'`IndexOutOfBoundsException` résultant de l'appel à `Heightmap#get()` pour obtenir l'altitude sur la zone du spawn. J'aurais aussi pu prendre une valeur par défaut si le (0, 0) est hors de la zone de la tuile générée, mais tant qu'à faire autant placer le spawn au centre de la tuile et puis ça règle la question

**Review by pyrollo — APPROVED** [review]

**Pull request merged**